python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -1279,18 +1278,28 @@ def _normalize_aggs(
         - A mapping of column name to aggs, such as
         {"a": ["sum"], "b": ["mean"]}, the aggs are applied to specified
         column.
+        - Pairs of column name and agg tuples passed as kwargs
+        eg. col1=("a", "sum"), col2=("b", "prod"). The output column names are
+        the keys. The aggs are applied to the corresponding column in the tuple.
         Each agg can be string or lambda functions.
         """
 
         aggs_per_column: Iterable[AggType | Iterable[AggType]]
-        if isinstance(aggs, dict):
-            column_names, aggs_per_column = aggs.keys(), aggs.values()
-            columns = tuple(self.obj._data[col] for col in column_names)
+        if aggs or isinstance(aggs, dict):
```

Comment:
If I understand correctly, you made this change to handle the fact that the legacy dask-cudf logic may be handing you an empty `dict` object here.

It makes sense that you are trying to handle this case. However, I feel that it would be better to raise the TypeError here for an empty aggregation spec. That is, I do think there is a bug in the legacy dask-cudf logic.

Instead of protecting against the empty dict here, can you try fixing the `CudfDataFrameGroupBy._make_groupby_method_aggs` definition (in `dask_cudf/groupby.py`) to look something like:

```python
    @_dask_cudf_performance_tracking
    def _make_groupby_method_aggs(self, agg_name):
        """Create aggs dictionary for aggregation methods"""
        # Leave out the columns we are grouping on unless we are
        # grouping by all columns (in which case we should include
        # all columns in the output dict)
        by = self.by.copy() if isinstance(self.by, list) else [self.by]
        target_cols = [c for c in self.obj.columns if c not in by] or by
        return {c: agg_name for c in target_cols}
```

Reply:
**EDIT**: I *think* we want something like the following in `dask_cudf`:

```python
    @_dask_cudf_performance_tracking
    def _make_groupby_method_aggs(self, agg_name):
        """Create aggs dictionary for aggregation methods"""
        # Leave out the columns we are grouping by. If we are
        # grouping by ALL columns, we cannot use a dict to
        # define the aggregation spec
        by = self.by.copy() if isinstance(self.by, list) else [self.by]
        target_cols = [c for c in self.obj.columns if c not in by]
        if not target_cols:
            return agg_name
        return {c: agg_name for c in target_cols}
```

This mess is all due to the fact that the legacy dask-cudf code path tries to represent **all** aggregations with a `dict` spec (even when there are no columns to aggregate). An alternative to this change is to simply keep your `dict` check in `cudf`, and add a note that the empty-dict handling can be removed when the legacy dask-cudf API is removed. I think I'm open to either solution.

Reply:
> If I understand correctly, you made this change to handle the fact that the legacy dask-cudf logic may be handing you an empty `dict` object here.

That's right.

Reply:
> This mess is all due to the fact that the legacy dask-cudf code path tries to represent **all** aggregations with a `dict` spec (even when there are no columns to aggregate). An alternative to this change is to simply keep your `dict` check in `cudf`, and add a note that the empty-dict handling can be removed when the legacy dask-cudf API is removed. I think I'm open to either solution.

Okay then I'll probably just keep the check and add the TODO. Im curious, when will the legacy dask_cudf API be removed @rjzamora ?

Reply:
>Im curious, when will the legacy dask_cudf API be removed

There is no concrete plan yet. My sense is that we should add a deprecate warning in 24.12 (**maybe** in 24.10 if Curator moves away from the legacy API soon). It's unclear to me how long we will need to wait for removal, but I'm excited to remove a lot of unnecessary code and tests once we do :)